### DISCUSSION TITLE
ruleset使用source时无法启动

### DISCUSSION THREAD

**Discussion by p-w-p** [immortalwrt/homeproxy#144] [Q&A] [answered]
如题，ruleset使用binary时可以运行，使用source时无法启动。
![image](https://github.com/user-attachments/assets/e22070f5-d0fb-417d-aa15-014e5af1de5d)
![image](https://github.com/user-attachments/assets/cccf6c72-9c7d-4c2b-a36b-0c2f69747519)
同时HomeProxy日志与sing-box 客户端日志都没有输出任何错误信息。
![image](https://github.com/user-attachments/assets/7533aebf-ef3b-47bc-9631-3dab47e27c71)
在ruleset使用binary启动的情况下，即使配置了DNS规则也无法正常生效，全部请求走默认DNS。
![image](https://github.com/user-attachments/assets/5eba0f37-4518-4a6d-9cd5-85d499cb8187)
麻烦各位大佬们帮忙看看:D

Cc @muink

**Comment by muink** [comment]
发下配置看看
source 接受 [source-format](https://sing-box.sagernet.org/configuration/rule-set/source-format/) 的 json文件
binary 接受 `sing-box rule-set compile [--output <file-name>.srs] <file-name>.json` 后的 srs文件

**Comment by p-w-p** [comment]
> 发下配置看看 source 接受 [source-format](https://sing-box.sagernet.org/configuration/rule-set/source-format/) 的 json文件 binary 接受 `sing-box rule-set compile [--output <file-name>.srs] <file-name>.json` 后的 srs文件

测试可以启动的binary
https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/geo/geosite/gfw.srs
测试无法启动的source
https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/geo/geosite/gfw.json

顺带提一下路由规则那边也用到了名称为sitegfw的规则集,那边工作正常但放到DNS规则里无法分流 XD

**Comment by muink** [comment]
这两个文件, 内容是一致的
我猜测单纯是 gfw.json 还没下载下来
进一步检查请提供配置文件
而且说实话, remote 不推荐用 json, 即便是压缩成单行的 json, 还是要比 srs 大一倍

**Comment by muink** [comment, marked as answer]
![image](https://github.com/user-attachments/assets/1e90c688-49e9-4c82-b852-8e679d6966a9)
outbound 用法错了, 删掉即可
outbound 什么情况下用, 参考这个 [issues](https://github.com/muink/luci-app-homeproxy/issues/4)

**Comment by p-w-p** [comment]
> ![image](https://private-user-images.githubusercontent.com/7929014/356609969-1e90c688-49e9-4c82-b852-8e679d6966a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyMTk5NjgsIm5iZiI6MTcyMzIxOTY2OCwicGF0aCI6Ii83OTI5MDE0LzM1NjYwOTk2OS0xZTkwYzY4OC00OWU5LTRjODItYjg1Mi04ZTY3OWQ2OTY2YTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDlUMTYwNzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWFmYmI5MzViNTQ3NGZlMWVlMGQ3NWVhNWY4MzU2MzE3NmNlNDkzMzNhMTZhMDc2YjhmNjNjOGQ2NTY2ODgwMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.NA8iy8hPZhsDEKMvniEDegq7KD1AICO77Q8QMGaGPDM) outbound 用法错了, 删掉即可 outbound 什么情况下用, 参考这个 [issues](https://github.com/muink/luci-app-homeproxy/issues/4)

确实是这个问题，感谢大佬😍
倒是还有个奇怪的问题，dns设置为tls://dns.google的时候会报ERROR dns: exchange failed for google.com. IN A: io: read/write on closed pipe，但改成tls://8.8.8.8的时候又正常。在使用tls://dns.google时地址解析器设置WAN下发或者AliDNS都不行，大佬能猜测下大概是什么原因不😭
dns.google会命中sitegfw这条ruleset，不知是否有影响？

**Comment by muink** [comment]
你建个8888的server
用于解析tls://dns.google的server试试看

**Comment by p-w-p** [comment]
> 你建个8888的server 用于解析tls://dns.google的server试试看

新建了个用于解析dns的上游并把dns.alidns.com和dns.google的都切了过去，发现上游那边只收到了dns.alidns.com的请求记录并没有dns.google的请求记录，难不成又是死循环🤦‍♂️

**Comment by muink** [comment]
![image](https://github.com/user-attachments/assets/a55a4f55-88a4-4573-93f6-d169015f59ec)
是在 hp 里建服务器, 让 boot 服务器去解析 带域名的服务器的域名

**Reply by p-w-p** [reply]
![image](https://github.com/user-attachments/assets/042f2e61-5545-4ba6-9675-3d8bcc4d1248)
![image](https://github.com/user-attachments/assets/a00ae60d-0c45-4205-8767-760ded359e23)
![image](https://github.com/user-attachments/assets/b4db4a3d-8a2a-4b6a-97a9-3db86e397e8f)
![image](https://github.com/user-attachments/assets/c523d04a-7708-4477-b689-1c6e36de0d5f)

**Reply by muink** [reply]
贴配置文件吧 /etc/config/homeproxy 算不了命
把节点地址脱敏就好

**Reply by muink** [reply]
叫你节点地址脱敏 不是叫你把 自己觉得不相干的 section 块都给删掉
你删了个 subscription, node 我页面都加载不出来
还得自己去补全参数

**Reply by p-w-p** [reply]
> 叫你节点地址脱敏 不是叫你把 自己觉得不相干的 section 块都给删掉 你删了个 subscription, node 我页面都加载不出来 还得自己去补全参数

sry

**Reply by muink** [reply]
给你修好了, 又是没写 `any-out` 的问题
我真的哭死
```
config homeproxy 'infra'
	option __warning 'DO NOT EDIT THIS SECTION, OR YOU ARE ON YOUR OWN!'
	option common_port '22,53,80,143,443,465,853,873,993,995,8080,8443,9418'
	option mixed_port '5330'
	option redirect_port '5331'
	option tproxy_port '5332'
	option dns_port '5333'
	option china_dns_port '5334'
	option tun_name 'singtun0'
	option tun_addr4 '172.19.0.1/30'
	option tun_addr6 'fdfe:dcba:9876::1/126'
	option tun_mtu '9000'
	option table_mark '100'
	option self_mark '100'
	option tproxy_mark '101'
	option tun_mark '102'

config homeproxy 'config'
	option routing_mode 'custom'
	option routing_port 'common'
	option proxy_mode 'redirect_tproxy'
	option ipv6_support '1'

config homeproxy 'control'
	option lan_proxy_mode 'disabled'

config homeproxy 'routing'
	option sniff_override '0'
	option default_outbound 'direct-out'
	option bypass_cn_traffic '1'

config homeproxy 'subscription'
	option auto_update '0'
	option allow_insecure '0'
	option update_via_proxy '0'
	option filter_nodes 'blacklist'
	option packet_encoding 'xudp'

config homeproxy 'dns'
	option dns_strategy 'prefer_ipv4'
	option default_server 'AliDNS'

config dns_server 'AliDNS'
	option label 'AliDNS'
	option enabled '1'
	option address 'udp://223.5.5.5'
	option outbound 'direct-out'

config dns_server 'AliDNS_DoT'
	option label 'AliDNS_DoT'
	option enabled '1'
	option address 'tls://dns.alidns.com'
	option address_resolver 'AliDNS'
	option outbound 'direct-out'

config dns_server 'Google'
	option label 'Google'
	option enabled '1'
	option address 'tcp://8.8.8.8'
	option outbound 'Node1'

config dns_server 'Google_DoT'
	option label 'Google_DoT'
	option enabled '1'
	option address 'tls://dns.google'
	option address_resolver 'Google'
	option outbound 'Node1'

config dns_rule 'Nodes_Domain'
	option label 'Nodes_Domain'
	option enabled '1'
	option mode 'default'
	list outbound 'any-out'
	option server 'AliDNS_DoT'

config dns_rule 'gfwsite'
	option label 'gfwsite'
	option enabled '1'
	option mode 'default'
	option server 'Google'
	list rule_set 'sitegfw'

config routing_node 'Node1'
	option label 'Node1'
	option enabled '1'
	option node 'd7376bc253c6e71aa20d95fc75521ea7'

config routing_rule 'gfw'
	option label 'gfw'
	option enabled '1'
	option mode 'default'
	list rule_set 'sitegfw'
	option outbound 'Node1'
	list protocol 'http'
	list protocol 'tls'
	list protocol 'quic'
	list protocol 'stun'

config routing_rule 'telegram'
	option label 'telegram'
	option enabled '1'
	option mode 'default'
	list rule_set 'iptelegram'
	option outbound 'Node1'

config ruleset 'sitegfw'
	option label 'sitegfw'
	option enabled '1'
	option type 'remote'
	option format 'binary'
	option url 'https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/geo/geosite/gfw.srs'
	option outbound 'Node1'

config ruleset 'iptelegram'
	option label 'iptelegram'
	option enabled '1'
	option type 'remote'
	option format 'binary'
	option url 'https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/geo/geoip/telegram.srs'
	option outbound 'Node1'

config node 'd7376bc253c6e71aa20d95fc75521ea7'
	option label 'SS-Node1'
	option type 'shadowsocks'
	option address 'aa.bb.cc'
	option port '12345'
	option shadowsocks_encrypt_method 'aes-256-gcm'
	option password '1234567'

```

**Reply by p-w-p** [reply]
> 给你修好了, 又是没写 `any-out` 的问题 我真的哭死
> 
> ```
> config homeproxy 'infra'
> 	option __warning 'DO NOT EDIT THIS SECTION, OR YOU ARE ON YOUR OWN!'
> 	option common_port '22,53,80,143,443,465,853,873,993,995,8080,8443,9418'
> 	option mixed_port '5330'
> 	option redirect_port '5331'
> 	option tproxy_port '5332'
> 	option dns_port '5333'
> 	option china_dns_port '5334'
> 	option tun_name 'singtun0'
> 	option tun_addr4 '172.19.0.1/30'
> 	option tun_addr6 'fdfe:dcba:9876::1/126'
> 	option tun_mtu '9000'
> 	option table_mark '100'
> 	option self_mark '100'
> 	option tproxy_mark '101'
> 	option tun_mark '102'
> 
> config homeproxy 'config'
> 	option routing_mode 'custom'
> 	option routing_port 'common'
> 	option proxy_mode 'redirect_tproxy'
> 	option ipv6_support '1'
> 
> config homeproxy 'control'
> 	option lan_proxy_mode 'disabled'
> 
> config homeproxy 'routing'
> 	option sniff_override '0'
> 	option default_outbound 'direct-out'
> 	option bypass_cn_traffic '1'
> 
> config homeproxy 'subscription'
> 	option auto_update '0'
> 	option allow_insecure '0'
> 	option update_via_proxy '0'
> 	option filter_nodes 'blacklist'
> 	option packet_encoding 'xudp'
> 
> config homeproxy 'dns'
> 	option dns_strategy 'prefer_ipv4'
> 	option default_server 'AliDNS'
> 
> config dns_server 'AliDNS'
> 	option label 'AliDNS'
> 	option enabled '1'
> 	option address 'udp://223.5.5.5'
> 	option outbound 'direct-out'
> 
> config dns_server 'AliDNS_DoT'
> 	option label 'AliDNS_DoT'
> 	option enabled '1'
> 	option address 'tls://dns.alidns.com'
> 	option address_resolver 'AliDNS'
> 	option outbound 'direct-out'
> 
> config dns_server 'Google'
> 	option label 'Google'
> 	option enabled '1'
> 	option address 'tcp://8.8.8.8'
> 	option outbound 'Node1'
> 
> config dns_server 'Google_DoT'
> 	option label 'Google_DoT'
> 	option enabled '1'
> 	option address 'tls://dns.google'
> 	option address_resolver 'Google'
> 	option outbound 'Node1'
> 
> config dns_rule 'Nodes_Domain'
> 	option label 'Nodes_Domain'
> 	option enabled '1'
> 	option mode 'default'
> 	list outbound 'any-out'
> 	option server 'AliDNS_DoT'
> 
> config dns_rule 'gfwsite'
> 	option label 'gfwsite'
> 	option enabled '1'
> 	option mode 'default'
> 	option server 'Google'
> 	list rule_set 'sitegfw'
> 
> config routing_node 'Node1'
> 	option label 'Node1'
> 	option enabled '1'
> 	option node 'd7376bc253c6e71aa20d95fc75521ea7'
> 
> config routing_rule 'gfw'
> 	option label 'gfw'
> 	option enabled '1'
> 	option mode 'default'
> 	list rule_set 'sitegfw'
> 	option outbound 'Node1'
> 	list protocol 'http'
> 	list protocol 'tls'
> 	list protocol 'quic'
> 	list protocol 'stun'
> 
> config routing_rule 'telegram'
> 	option label 'telegram'
> 	option enabled '1'
> 	option mode 'default'
> 	list rule_set 'iptelegram'
> 	option outbound 'Node1'
> 
> config ruleset 'sitegfw'
> 	option label 'sitegfw'
> 	option enabled '1'
> 	option type 'remote'
> 	option format 'binary'
> 	option url 'https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/geo/geosite/gfw.srs'
> 	option outbound 'Node1'
> 
> config ruleset 'iptelegram'
> 	option label 'iptelegram'
> 	option enabled '1'
> 	option type 'remote'
> 	option format 'binary'
> 	option url 'https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/geo/geoip/telegram.srs'
> 	option outbound 'Node1'
> 
> config node 'd7376bc253c6e71aa20d95fc75521ea7'
> 	option label 'SS-Node1'
> 	option type 'shadowsocks'
> 	option address 'aa.bb.cc'
> 	option port '12345'
> 	option shadowsocks_encrypt_method 'aes-256-gcm'
> 	option password '1234567'
> ```

嘶，我看看什么情况，不过我这边节点全部都是ip没使用domain喔？报DNS timeout的时候tg没问题，节点是能连接上的

**Reply by muink** [reply]
目前的故障是什么
我这边干净虚拟机的测试结果是已经可以正常访问google youtube了

**Reply by p-w-p** [reply]
> 目前的故障是什么 我这边干净虚拟机的测试结果是已经可以正常访问google youtube了

就算给dns.google和dns.alidns.com的option address_resolver指定了同一上游，上游有dns.alidns.com的查询日志但没有dns.google的查询日志。感觉很奇怪

**Reply by muink** [reply]
你的配置我已经查不出任何问题了
你可以自行检查 hp 的dns是否有问题 `dig www.youtube.com -p 5333`
如果没问题, 那就是 dnsmasq 设置了转发
检查 `/var/dnsmasq.d/`
检查 `http://192.168.100.1/cgi-bin/luci/admin/network/dhcp` -> `Forwards` -> `DNS Forwards`
检查 `http://192.168.100.1/cgi-bin/luci/admin/network/dhcp` -> `Resolv & Hosts Files` -> `Strict order`
![image](https://github.com/user-attachments/assets/f1d3b780-a3c7-46e5-916c-009fa333f9a4)

**Reply by p-w-p** [reply]
> 你的配置我已经查不出任何问题了 你可以自行检查 hp 的dns是否有问题 `dig www.youtube.com -p 5333` 如果没问题, 那就是 dnsmasq 设置了转发 检查 `/var/dnsmasq.d/` 检查 `http://192.168.100.1/cgi-bin/luci/admin/network/dhcp` -> `Forwards` -> `DNS Forwards` 检查 `http://192.168.100.1/cgi-bin/luci/admin/network/dhcp` -> `Resolv & Hosts Files` -> `Strict order` ![image](https://private-user-images.githubusercontent.com/7929014/356796115-f1d3b780-a3c7-46e5-916c-009fa333f9a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyNzUyNTIsIm5iZiI6MTcyMzI3NDk1MiwicGF0aCI6Ii83OTI5MDE0LzM1Njc5NjExNS1mMWQzYjc4MC1hM2M3LTQ2ZTUtOTE2Yy0wMDlmYTMzM2Y5YTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTBUMDcyOTEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2ExYThiNjQzODBmM2Q3NjNkMWNhMWRkYzNiMjgzMjdhODk2MjUzYjgxMDY2ZWFjZjcxYjkyMWFhY2MyN2NlMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ._3lW381uQ_5ae-qudNkmHOIS600xKMSPltP__pu9swc)

好的，十分感谢！

**Comment by p-w-p** [comment]
关于dns的部分似乎是由于客户端与服务端混用导致的兼容性问题，更换服务端core后问题消失。
